src/zon.zig

```diff
@@ -0,0 +1,831 @@
+                return self.fail(
+                    .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+                    "float value '{}' cannot be stored in integer type '{}'",
+                    .{ val, res_ty.fmt(self.sema.pt) },
+                );
+            }
+
+            return self.sema.pt.zcu.intern_pool.get(gpa, self.sema.pt.tid, .{
+                .int = .{
+                    .ty = res_ty.toIntern(),
+                    .storage = .{ .big_int = rational.p.toConst() },
+                },
+            });
+        },
+        .char_literal => |val| {
+            const rhs: u32 = val;
+            // If our result is a fixed size integer, check that our value is not out of bounds
+            if (Type.zigTypeTag(res_ty, self.sema.pt.zcu) == .int) {
+                const lhs_info = res_ty.intInfo(self.sema.pt.zcu);
+                // If lhs has less than 64 bits, we bounds check. We check at 64 instead of 32 in
+                // case LHS is signed.
+                if (std.math.cast(u6, lhs_info.bits)) |bits| {
+                    const max_int: i64 = if (bits == 0) b: {
+                        break :b 0;
+                    } else b: {
+                        break :b (@as(i64, 1) << (bits - @intFromBool(lhs_info.signedness == .signed))) - 1;
+                    };
+                    if (rhs > max_int) {
+                        return self.fail(
+                            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+                            "type '{}' cannot represent integer value '{}'",
+                            .{ res_ty.fmt(self.sema.pt), rhs },
+                        );
+                    }
+                }
+            }
+            return self.sema.pt.zcu.intern_pool.get(gpa, self.sema.pt.tid, .{
+                .int = .{
+                    .ty = res_ty.toIntern(),
+                    .storage = .{ .i64 = rhs },
+                },
+            });
+        },
+
+        else => return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "expected type '{}'",
+            .{res_ty.fmt(self.sema.pt)},
+        ),
+    };
+}
+
+fn lowerFloat(
+    self: LowerZon,
+    node: Zoir.Node.Index,
+    res_ty: Type,
+) !InternPool.Index {
+    @setFloatMode(.strict);
+    switch (node.get(self.file.zoir.?)) {
+        .int_literal => |int| switch (int) {
+            .small => |val| return self.sema.pt.intern(.{ .float = .{
+                .ty = res_ty.toIntern(),
+                .storage = switch (res_ty.toIntern()) {
+                    .f16_type => .{ .f16 = @floatFromInt(val) },
+                    .f32_type => .{ .f32 = @floatFromInt(val) },
+                    .f64_type => .{ .f64 = @floatFromInt(val) },
+                    .f80_type => .{ .f80 = @floatFromInt(val) },
+                    .f128_type, .comptime_float_type => .{ .f128 = @floatFromInt(val) },
```

Comment:
All of these `switch`es in this function are invalid, because `c_longdouble` exists, and the size of that one is variable. See `Zcu.PerThread.floatValue` for inspiration -- in fact, just *call* `Zcu.PerThread.floatValue` :P